Index: backports-5.15.74-1/include/uapi/linux/nl80211.h
===================================================================
--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -3182,7 +3182,7 @@ enum nl80211_attrs {
 #define NL80211_HE_MIN_CAPABILITY_LEN           16
 #define NL80211_HE_MAX_CAPABILITY_LEN           54
 #define NL80211_MAX_NR_CIPHER_SUITES		5
-#define NL80211_MAX_NR_AKM_SUITES		2
+#define NL80211_MAX_NR_AKM_SUITES		10
 
 #define NL80211_MIN_REMAIN_ON_CHANNEL_TIME	10
 
