From d7337c25e774ed96fc8f6d1390321bfc2dcac545 Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <meichia.chiu@mediatek.com>
Date: Tue, 1 Aug 2023 16:02:28 +0800
Subject: [PATCH 55/98] wifi: mt76: mt7996: Establish BA in VO queue

---
 mt7996/mac.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index 37cc94e..b24f237 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -1017,8 +1017,6 @@ mt7996_tx_check_aggr(struct ieee80211_sta *sta, struct sk_buff *skb)
 		return;
 
 	tid = skb->priority & IEEE80211_QOS_CTL_TID_MASK;
-	if (tid >= 6) /* skip VO queue */
-		return;
 
 	if (is_8023)
 		fc = IEEE80211_FTYPE_DATA |
-- 
2.18.0

