Index: hostapd-2022-07-29-b704dc72/src/common/wpa_common.c
===================================================================
--- hostapd-2022-07-29-b704dc72.orig/src/common/wpa_common.c
+++ hostapd-2022-07-29-b704dc72/src/common/wpa_common.c
@@ -957,12 +957,30 @@ int wpa_ft_mic(int key_mgmt, const u8 *k
 			link_rsne_len[ml_info.link_id] = rsnie_len;
 			link_rsnxe[ml_info.link_id] = rsnxe;
 			link_rsnxe_len[ml_info.link_id] = rsnxe_len;
+			wpa_printf(MSG_INFO, "FT: ml_info.link_id = %d",
+					ml_info.link_id);
 
 			if (!tx_reassoc_rsp) {
 				for (i = 0; i < ml_info.prof_num; i++) {
 					struct ieee802_11_elems elems;
 					u8 link_id = ml_info.profiles[i].link_id;
 
+					wpa_printf(MSG_INFO, "FT: link_id = %d",
+						link_id);
+
+					if (link_id == ml_info.link_id) {
+						wpa_printf(MSG_INFO, "FT: ml_info.link_id is equal to per-sta profile link id");
+						ml_info.link_id++;
+
+						link_addr[ml_info.link_id] = main_link_addr;
+						link_rsne[ml_info.link_id] = rsnie;
+						link_rsne_len[ml_info.link_id] = rsnie_len;
+						link_rsnxe[ml_info.link_id] = rsnxe;
+						link_rsnxe_len[ml_info.link_id] = rsnxe_len;
+
+						valid_links |= BIT(ml_info.link_id);
+					}
+
 					valid_links |= BIT(link_id);
 					link_addr[link_id] = ml_info.profiles[i].addr;
 					link_rsne[link_id] = rsnie;
@@ -1198,7 +1216,7 @@ int wpa_ft_mic(int key_mgmt, const u8 *k
 		res = 0;
 	}
 #endif /* CONFIG_SHA384 */
-	if (kck_len == 16 && key_mgmt == WPA_KEY_MGMT_FT_SAE_EXT_KEY) {
+	if (kck_len == 16 && (key_mgmt & WPA_KEY_MGMT_FT_SAE_EXT_KEY)) {
 		u8 hash[SHA256_MAC_LEN];
 
 		if (hmac_sha256_vector(kck, kck_len, num_elem, addr, len, hash))
@@ -1206,7 +1224,7 @@ int wpa_ft_mic(int key_mgmt, const u8 *k
 		os_memcpy(mic, hash, 16);
 		res = 0;
 	}
-	if (kck_len == 16 && key_mgmt != WPA_KEY_MGMT_FT_SAE_EXT_KEY &&
+	if (kck_len == 16 && !(key_mgmt & WPA_KEY_MGMT_FT_SAE_EXT_KEY) &&
 	    omac1_aes_128_vector(kck, num_elem, addr, len, mic) == 0)
 		res = 0;
 
