From 5d04c65f6a625dea4b8ff7cfa35311dbfa2e4ae7 Mon Sep 17 00:00:00 2001
From: Evelyn Tsai <evelyn.tsai@mediatek.com>
Date: Wed, 30 Aug 2023 04:23:37 +0800
Subject: [PATCH 1002/1005] hostapd: mtk: ucode: add support for ucode to parse
 BW320MHz info

---
 src/utils/ucode.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/utils/ucode.c b/src/utils/ucode.c
index 2beeb9a..122c619 100644
--- a/src/utils/ucode.c
+++ b/src/utils/ucode.c
@@ -85,6 +85,10 @@ uc_value_t *uc_wpa_freq_info(uc_vm_t *vm, size_t nargs)
 	case 2:
 		chanwidth = CONF_OPER_CHWIDTH_160MHZ;
 		break;
+	case 9:
+		width = 3;
+		chanwidth = CONF_OPER_CHWIDTH_320MHZ;
+		break;
 	default:
 		return NULL;
 	}
-- 
2.18.0

